tests/testthat/test-finalsize_grps_cpp_newton.R

```diff
@@ -0,0 +1,218 @@
+  psusc <- matrix(
+    data = 1, nrow = n_demo_grps, ncol = n_risk_grps
+  )
+  psusc <- t(apply(psusc, 1, \(x) x / sum(x)))
+  susc <- matrix(
+    data = 1, nrow = n_demo_grps, ncol = n_risk_grps
+  )
+
+  # prepare control
+  control <- list(
+    iterations = 10000,
+    tolerance = 1e-6,
+    step_rate = 1.9,
+    adapt_step = TRUE
+  )
+
+  epi_outcome <- final_size_grps_cpp(
+    contact_matrix = r0 * c_matrix,
+    demography_vector = d_vector,
+    p_susceptibility = psusc,
+    susceptibility = susc,
+    solver = "newton",
+    control = control
+  )
+
+  expect_type(
+    epi_outcome, "double"
+  )
+  # check that values are not NaN
+  expect_true(
+    all(!is.nan(epi_outcome))
+  )
+  # check that solver returns no nas
+  expect_false(
+    any(is.na(epi_outcome))
+  )
+  # check that solver returns no inf
+  expect_false(
+    any(is.infinite(epi_outcome))
+  )
+  # check that solver returns values within range
+  expect_true(
+    all(epi_outcome > 0)
+  )
+  expect_true(
+    all(epi_outcome < 1)
+  )
+  # check for size of the vector
+  expect_equal(
+    n_demo_grps,
+    length(epi_outcome)
+  )
+})
+
+# Check sensitivity to susceptibility using newton solver
+test_that("Check that more susceptible demo-grps have higher final size", {
+  r0 <- 1.3
+  polymod <- socialmixr::polymod
+  contact_data <- socialmixr::contact_matrix(
+    polymod,
+    countries = "United Kingdom",
+    age.limits = c(0, 20, 40),
+    split = TRUE
+  )
+  c_matrix <- t(contact_data$matrix)
+  d_vector <- contact_data$participants$proportion
+
+  # Scale contact matrix to demography
+  c_matrix <- apply(
+    c_matrix, 1, function(r) r / d_vector
+  )
+
+  n_demo_grps <- length(d_vector)
+  n_risk_grps <- 4
+
+  # prepare p_susceptibility and susceptibility
+  # susceptibility of demography group i = 1 is 10x i = 2, 3
+  psusc <- matrix(
+    data = 1, nrow = n_demo_grps, ncol = n_risk_grps
+  )
+  psusc <- t(apply(psusc, 1, \(x) x / sum(x)))
+  susc <- rbind(
+    rep(1, n_risk_grps),
+    rep(0.1, n_risk_grps),
+    rep(0.1, n_risk_grps)
+  )
+
+  # prepare control
+  control <- list(
+    iterations = 10000,
+    tolerance = 1e-6,
+    step_rate = 1.9,
+    adapt_step = TRUE
+  )
+
+  epi_outcome <- final_size_grps_cpp(
+    contact_matrix = r0 * c_matrix,
+    demography_vector = d_vector,
+    p_susceptibility = psusc,
+    susceptibility = susc,
+    solver = "newton",
+    control = control
+  )
+
+  expect_vector(
+    epi_outcome,
+    ptype = numeric()
+  )
+  # check that solver returns values within range
+  expect_true(
+    all(epi_outcome > 0)
+  )
+  expect_true(
+    all(epi_outcome < 1)
+  )
+  # check for size of the vector
+  expect_equal(
+    n_demo_grps,
+    length(epi_outcome)
+  )
+  # check that first group has higher final size
+  expect_gt(
+    epi_outcome[1], epi_outcome[n_demo_grps]
+  )
+})
+
+# check for correct final size calculation in complex data case
+# using newton solver
```

Comment:
Do we have a case, e.g. a published paper, with known values of final size? This would be even better than checking the range.

You now added tests to ensure all implementations return the same result but what if the result was wrong in all cases? Do we have a reference implementation / set of values?

Reply:
I'll look around, and also ask one of the PIs. My understanding is that the tests for 'correct answers' in the files `test-*_solver.R`, as well as those using the `upper_limit` function in the files `test-*_solver_vary_r0` ensure that the answers are exactly correct in some simple cases at least.